File: tests/deepflow-typescript/src/services/report.service.ts
/**
 * Report service - HOP 3: Business logic for reports.
 *
 * Handles report generation with XSS vulnerability.
 */

import { DataTransformer } from '../processors/data.transformer';

export class ReportService {
  private transformer: DataTransformer;

  constructor() {
    this.transformer = new DataTransformer();
  }

  /**
   * Generate report with custom title.
   *
   * HOP 3: Passes tainted title to processor.
   *
   * @param title - TAINTED title - XSS vector
   * @param data - Report data
   */
  async generate(title: string, data: any): Promise<any> {
    return this.transformer.prepareReportGeneration(title, data);
  }

  /**
   * Preview report content.
   *
   * @param content - TAINTED content - XSS vector
   */
  async preview(content: string): Promise<string> {
    return this.transformer.preparePreview(content);
  }

  /**
   * Search reports.
   *
   * @param query - TAINTED query - SQL injection vector
   */
  async search(query: string): Promise<any> {
    return this.transformer.prepareReportSearch(query);
  }
}

File: tests/deepflow-python/app/services/report_service.py
"""Report service - HOP 3: Business logic for reports.

Handles report generation with multiple tainted inputs.
"""

from app.processors.data_transformer import DataTransformer


class ReportService:
    """Report business logic service.

    HOP 3: Receives multiple tainted inputs (title, format, callback_url)
    and routes them to appropriate processors.
    """

    def __init__(self):
        self.transformer = DataTransformer()

    async def generate(
        self,
        title: str,
        output_format: str,
        callback_url: str | None = None,
    ) -> dict:
        """Generate a report.

        HOP 3: Multiple tainted inputs flow through:
        - title -> XSS chain (14 hops)
        - output_format -> Command Injection chain (12 hops)
        - callback_url -> SSRF chain (8 hops)

        Args:
            title: TAINTED - flows to template rendering
            output_format: TAINTED - flows to command execution
            callback_url: TAINTED - flows to HTTP request
        """
        # Each parameter flows to different vulnerability chains
        report_data = self.transformer.prepare_report(
            title=title,
            output_format=output_format,
            callback_url=callback_url,
        )
        return report_data

    async def export(self, report_id: str, format: str) -> dict:
        """Export report in specified format.

        Args:
            report_id: Report identifier
            format: TAINTED - flows to command execution
        """
        return self.transformer.prepare_export(report_id, format)
